compiler/scenario-service/server/src/main/scala/com/digitalasset/daml/lf/scenario/Conversions.scala

```diff
@@ -88,58 +89,87 @@ final class Conversions(
+            //     errors so we never produce ContractNotFound
+            builder.setCrash(s"contract ${cid.coid} not found")
+          case TemplatePreconditionViolated(tid, optLoc, arg) =>
+            val uepvBuilder = proto.ScenarioError.TemplatePreconditionViolated.newBuilder
+            optLoc.map(convertLocation).foreach(uepvBuilder.setLocation)
+            builder.setTemplatePrecondViolated(
+              uepvBuilder
+                .setTemplateId(convertIdentifier(tid))
+                .setArg(convertValue(arg))
+                .build
+            )
+          case ContractNotActive(coid, tid, consumedBy) =>
+            builder.setUpdateLocalContractNotActive(
+              proto.ScenarioError.ContractNotActive.newBuilder
+                .setContractRef(mkContractRef(coid, tid))
+                .setConsumedBy(proto.NodeId.newBuilder.setId(consumedBy.toString).build)
+                .build
+            )
+          case LocalContractKeyNotVisible(coid, gk, actAs, readAs, stakeholders) =>
+            builder.setScenarioContractKeyNotVisible(
+              proto.ScenarioError.ContractKeyNotVisible.newBuilder
+                .setContractRef(mkContractRef(coid, gk.templateId))
+                .addAllActAs(actAs.map(convertParty(_)).asJava)
+                .addAllReadAs(readAs.map(convertParty(_)).asJava)
+                .addAllStakeholders(stakeholders.map(convertParty).asJava)
+                .build
+            )
+          case ContractKeyNotFound(gk) =>
+            builder.setScenarioContractKeyNotFound(
+              proto.ScenarioError.ContractKeyNotFound.newBuilder
+                .setTemplateId(convertIdentifier(gk.templateId))
+                .setKey(convertValue(gk.key))
+                .build
+            )
+          case DuplicateContractKey(key) =>
+            builder.setScenarioCommitError(
+              proto.CommitError.newBuilder.setUniqueKeyViolation(convertGlobalKey(key)).build
+            )
+          case CreateEmptyContractKeyMaintainers(tid, arg, key) =>
+            builder.setCreateEmptyContractKeyMaintainers(
+              proto.ScenarioError.CreateEmptyContractKeyMaintainers.newBuilder
+                .setArg(convertValue(arg))
+                .setTemplateId(convertIdentifier(tid))
+                .setKey(convertValue(key))
+            )
+          case FetchEmptyContractKeyMaintainers(tid, key) =>
+            builder.setFetchEmptyContractKeyMaintainers(
+              proto.ScenarioError.FetchEmptyContractKeyMaintainers.newBuilder
+                .setTemplateId(convertIdentifier(tid))
+                .setKey(convertValue(key))
+            )
+          case wtc: WronglyTypedContract =>
+            // TODO https://github.com/digital-asset/daml/issues/9974
+            //   (MK) This isn’t actually true. You can easily
+            //   coerceContractId your way into this error. Not for
+            //   this PR though.
```

Comment:
Added this to https://github.com/digital-asset/daml/issues/9974 's description, feel free to delete [the comment](https://github.com/digital-asset/daml/issues/9974#issuecomment-867649444) if unnecessary. 🙂